### PR TITLE
Add SNI support in Identity SDK

### LINKED ISSUE #10370: Support SubjectName/Issuer (SNI) certificate authentication for Service Principals

### PR DESCRIPTION
**Pull request by g2vinay**
Updates MSAL4J dependency to 1.7.1.
Accommodates the TenantProfiles added to `IAccount` in MSAL4J.
Adds Chained Certificate support to `ClientCertificateCredential`.
Adds option for user to configure sending x5c on `ClientCertificateCredentialBuilder`

Fixes #10370

### CODE REVIEW

File: sdk/identity/azure-identity/pom.xml [hunk 1]
```diff
@@ -32,7 +32,7 @@
     <dependency>
       <groupId>com.microsoft.azure</groupId>
       <artifactId>msal4j</artifactId>
-      <version>1.6.2</version> <!-- {x-version-update;com.microsoft.azure:msal4j;external_dependency} -->
+      <version>1.7.0</version> <!-- {x-version-update;com.microsoft.azure:msal4j;external_dependency} -->
```
Comment (g2vinay) [inline sdk/identity/azure-identity/pom.xml:35]: NOTE: I'll update the external_dep file before merging, as that will impact other libraries using MSAL too.

**Review by jianghaolu — APPROVED** [review]

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/InteractiveBrowserCredential.java [hunk 2]
```diff
@@ -119,8 +119,9 @@
     private AccessToken updateCache(MsalToken msalToken) {
         cachedToken.set(
                 new MsalAuthenticationAccount(
-                        new AuthenticationRecord(msalToken.getAuthenticationResult(),
-                                identityClient.getTenantId(), identityClient.getClientId())));
+                    new AuthenticationRecord(msalToken.getAuthenticationResult(),
+                                identityClient.getTenantId(), identityClient.getClientId()),
+                    msalToken.getAccount().getTenantProfiles()));
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/InteractiveBrowserCredential.java:124]: nit: The formatting looks a bit odd. Probably moving the instantiation of `AuthenticationRecord` outside of `set()` would make it more readable too.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 3]
```diff
@@ -172,9 +174,15 @@ private ConfidentialClientApplication getConfidentialClientApplication() {
...
-                        CertificateUtil.publicKeyFromPem(pemCertificateBytes));
+                    List<X509Certificate> x509CertificateList =  CertificateUtil.publicKeyFromPem(pemCertificateBytes);
+                    PrivateKey privateKey = CertificateUtil.privateKeyFromPem(pemCertificateBytes);
+                    if (x509CertificateList.size() == 1) {
+                        credential = ClientCredentialFactory.createFromCertificate(
+                            privateKey, x509CertificateList.get(0));
+                    } else {
+                        credential = ClientCredentialFactory.createFromCertificateChain(
+                            privateKey, x509CertificateList);
+                    }
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:185]: Do we need to special-case this for a list with size 1? Can `createFromCertificateChain` handle a list that has just 1 entry?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/CertificateUtil.java [hunk 4]
```diff
@@ -54,24 +56,31 @@ public static PrivateKey privateKeyFromPem(byte[] pem) {
...
-     * Extracts the X509Certificate certificate from a PEM certificate.
+     * Extracts the X509Certificate certificate/certificate-chain from a PEM certificate.
      * @param pem the contents of a PEM certificate.
-     * @return the X509Certificate certificate
+     * @return the {@link List} of X509Certificate certificate
      */
-    public static X509Certificate publicKeyFromPem(byte[] pem) {
-        Pattern pattern = Pattern.compile("(?s)-----BEGIN CERTIFICATE-----.*-----END CERTIFICATE-----");
+    public static List<X509Certificate> publicKeyFromPem(byte[] pem) {
+        Pattern pattern = Pattern.compile("(?s)-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----");
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/CertificateUtil.java:64]: Pattern compilation is expensive. So, maybe it's good to just make this pattern into a static final constant.

**Review by srnagar — APPROVED** [review]

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/CertificateUtil.java (hunk 4 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/CertificateUtil.java:64]: In this case, it is only called once during client construction and not in any API call.
So, having it localized will help to get it garbage collected, after use.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java (hunk 3 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:185]: yeah, we need to special case it, to ensure the cert chain is empty and not sent as part of the request, when there's only one cert to send.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/InteractiveBrowserCredential.java (hunk 2 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/InteractiveBrowserCredential.java:124]: agreed, 
will update it.

**Review by schaabs — APPROVED** [review]

**Pull request merged**